File: selectionSort.java
class Personagem {

	private String file;
	private String nome;
	private int altura; 
	private double peso;
	private String corDoCabelo;
	private String corDaPele;
	private String corDosOlhos;
	private String anoNascimento; 
	private String genero;
	private String homeworld;

	public Personagem()
	{
	}

	public Personagem(String nome)
	{
		this.nome = nome;
	}

	public Personagem clone( Personagem personagem)
	{
		Personagem personagem2 = new Personagem();
		personagem2.file = personagem.file;
		personagem2.nome = personagem.nome;
		personagem2.altura = personagem.altura;
		personagem2.peso = personagem.peso;
		personagem2.corDoCabelo = personagem.corDoCabelo;
		personagem2.corDaPele = personagem.corDaPele;
		personagem2.corDosOlhos = personagem.corDosOlhos;
		personagem2.anoNascimento = personagem.anoNascimento;
		personagem2.genero = personagem.genero;
		personagem2.homeworld = personagem.homeworld;
		return personagem2;
	}

	public void lerArquivo(String filename)
	{
		Arq.openRead(filename,"iso-8859-1");
		String linha = Arq.readLine();
		Arq.close();
		file = linha;

	}

	public void setNome()
	{
		String[] temp = file.split(",");
		String[] tmp = temp[0].split(":");
		if(tmp[1].contains("'"))
		{
			nome = tmp[1].replace("'","").trim();
		}
	}

	public String getNome()
	{
		return nome;
	}

	public void setAltura()
	{
		String[] temp = file.split("height':");
		String[] tmp = temp[1].split(",");
		String tp = tmp[0].replace("'","");
		tp = tp.replace(" ","");

		if(tp.contains("unknown")) 
		{	
			altura = 0;
		}
		else{
			altura = Integer.parseInt(tp);
		}
	}

	public int getAltura()
	{
		return altura;
	}

	public void setPeso()
	{
		String[] temp = file.split("mass':");
		String[] tmp = temp[1].split(",");
		String weight = tmp[0].replace("'","").trim();

		if (weight.contains("unknown"))
		{
			peso = 0;
		}
		else{
			peso = Double.parseDouble(weight);
		}
	}

	public double getPeso()
	{
		return peso;
	}

	public void setCorDoCabelo()
	{
		String[] temp = file.split("hair_color':");
		String[] tmp = temp[1].split("',");
		corDoCabelo = tmp[0].replace("'","").trim();
	}

	public String getCorDoCabelo()
	{
		return corDoCabelo;
	}

	public void setCorDaPele()
	{
		String[] temp = file.split("skin_color':");
		String[] tmp = temp[1].split("',");
		tmp[0] = tmp[0].replace("'","");
		corDaPele = tmp[0].trim();
	}

	public String getCorDaPele()
	{
		return corDaPele;
	}

	public void setCorDosOlhos()
	{
		String[] temp = file.split("eye_color':");
		String[] tmp = temp[1].split("',");
		corDosOlhos = tmp[0].replace("'","").trim();
	}

	public String getCorDosOlhos()
	{
		return corDosOlhos;
	}

	public void setAnoNascimento()
	{
		String[] temp = file.split("birth_year':");
		String[] tmp = temp[1].split(",");
		anoNascimento = tmp[0].replace("'","").trim();
	}

	public String getAnoNascimento()
	{
		return anoNascimento;
	}

	public void setGenero()
	{
		String[] temp = file.split("gender':");
		String[] tmp = temp[1].split(",");
		genero = tmp[0].replace("'","").trim();
	}

	public String getGenero()
	{
		return genero;
	}

	public void setHomeworld()
	{
		String[] temp = file.split("homeworld':");
		String[] tmp = temp[1].split(",");
		homeworld = tmp[0].replace("'","").trim();
	}

	public String getHomeworld()
	{
		return homeworld;
	}

	public void setAll()
	{
		setNome();
		setAltura();
		setPeso();
		setCorDoCabelo();
		setCorDaPele();
		setCorDosOlhos();
		setAnoNascimento();
		setGenero();
		setHomeworld();

	}

	public void imprimir()
	{
		MyIO.print(" ## "+nome);
		MyIO.print(" ## "+altura);

		if(String.valueOf(peso).contains(".0"))
		{
			MyIO.print(" ## "+String.valueOf(peso).replace(".0",""));
		}
		else{
			MyIO.print(" ## "+peso);
		}
		MyIO.print(" ## "+corDoCabelo);
		MyIO.print(" ## "+corDaPele);
		MyIO.print(" ## "+corDosOlhos);
		MyIO.print(" ## "+anoNascimento);
		MyIO.print(" ## "+genero);
		MyIO.print(" ## "+homeworld+" ## "+"\n");
	}
}

//=========================================================FIM DA CLASSE PERSONAGEM=========================================================================

public class selectionSort{

	private Personagem[] array;
	private int quantidade;
	private int tamanho;
	public int mov;
	public int comp;

	selectionSort(){
	}

	selectionSort(int tamanho)
	{
		array = new Personagem[tamanho];
		quantidade = 0;

		for(int x = 0; x < array.length; x++)
		{
			array[x] = new Personagem();
		}
	}

	public void construirLista(String[] linha,int numEntrada)
	{
		for(int x = 0; x < numEntrada; x++)
		{
			array[x].lerArquivo(linha[x]);
			array[x].setAll();
			quantidade++;
		}
	}

	public void inserirInicio(Personagem personagem)
	{
		if(quantidade < array.length)
		{
			for(int x = quantidade; x > 0; x--)
			{
				array[x] = array[x].clone(array[x-1]);
			}
			array[0] = array[0].clone(personagem);
			quantidade++;
		}
		else{
			MyIO.println("O array esta cheio");
		}
	}

	public void inserir(Personagem personagem, int pos)
	{
		if(quantidade < array.length)
		{
			for(int x = quantidade; x > pos; x--)
			{
				array[x] = array[x].clone(array[x-1]);
			}
			array[pos] = array[pos].clone(personagem);
			quantidade++;
		}
		else{
			MyIO.println("O array esta cheio");
		}
	}

	public void inserirFim(Personagem personagem)
	{
		if(quantidade < array.length)
		{
			array[quantidade] = array[quantidade].clone(personagem);
			quantidade++;
		}
		else{
			MyIO.println("O array esta cheio");
		}
	}

	public Personagem removerInicio()
	{
		Personagem resp = array[0].clone(array[0]);
		resp.setNome();
		if(quantidade != 0)
		{
			quantidade--;

			for(int x = 0; x < quantidade; x++)
			{
				array[x] = array[x].clone(array[x+1]);
			}
		}
		return resp;
	}

	public Personagem remover(int pos)
	{
		Personagem resp = array[pos].clone(array[pos]);

		if (quantidade > 0 && pos >= 0 && pos < quantidade)
		{
			quantidade--;

			if(pos == quantidade)
			{
				removerFim();
			}
			else if(pos == 0)
			{
				removerInicio();
			}
			else{
				for(int x = pos; x < quantidade; x++)
				{
					array[x] = array[x+1];
				}
				resp.setNome();
			}
		}
		return resp;
	}

	public Personagem removerFim()
	{
		quantidade--;
		array[quantidade].setNome();
		return array[quantidade];
	}

	public void sort()
	{
		for(int i = 0; i < quantidade; i++)
		{
			int menor = i;
			for(int j = (i+1); j < quantidade; j++)
			{
				if(array[j].getNome().compareTo(array[menor].getNome( )) < 0 )
				{
					comp++;
					menor = j;
				}
			}
			Personagem tmp = new Personagem();
			tmp = tmp.clone(array[menor]);
			mov++;
			array[menor] = array[menor].clone(array[i]);
			mov++;
			array[i] = array[i].clone(tmp);
			mov++;
		}
	}

	public void mostrarArray()
	{
		for(int x = 0; x < quantidade; x++)
		{
			array[x].setAll();
		}
		for(int y = 0; y < quantidade; y++)
		{
			//MyIO.print("["+y+"] ");
			array[y].imprimir();
		}
	}

	public static boolean isFim(String s){
		return (s.length() >= 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M');
	}

	public static void main (String[] Args)
	{
		MyIO.setCharset("utf-8");
		String[] linha = new String[100];
		selectionSort lista = new selectionSort(100);
		Personagem personagem = new Personagem();
		int numEntrada = 0;

		do {
			linha[numEntrada] = MyIO.readLine();
		} while (isFim(linha[numEntrada++]) == false);
		numEntrada--;

		lista.tamanho = numEntrada;

		long startTime = System.nanoTime();

		for(int x = 0; x < numEntrada; x++)
		{
			personagem.lerArquivo(linha[x]);
			personagem.setAll();
			lista.inserirInicio(personagem);
		}
		lista.sort();
		lista.mostrarArray();

		long endTime = System.nanoTime();
		long execTime= (endTime - startTime);
	
		Arq.openWriteClose("668007_selecao.txt", "UTF-8", "Matricula: 668007\t"+"Tempo de execucao:\t"+execTime+"\t"+"Movimentacoes: \t"+lista.mov+"\tComparacoes:\t"+lista.comp);
	}
}

File: selectionSort.cc
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define bool      short
#define true      1
#define false     0
#define equals(a, b) (((strcmp(a, b) == 0) ? true : false))
#define NUMENTRADA 1000
#define TAMLINHA   1000

typedef struct personagem{ //Struct utilizado para guardar as informacoes relevantes
	char* file;
	char* nome;
	int altura;
	double peso;
	char* corDoCabelo;
	char* corDaPele;
	char* corDosOlhos;
	char* anoNascimento;
	char* genero;
	char* homeworld;   
}PERSONAGEM;

PERSONAGEM newPersonagem()
{
	PERSONAGEM novo;
	novo.nome = (char*)malloc(sizeof(char*)*1000);
	novo.altura = 0;
	novo.peso = 0;
	novo.corDoCabelo = (char*)malloc(sizeof(char*)*1000);
	novo.corDaPele = (char*)malloc(sizeof(char*)*1000);
	novo.corDosOlhos = (char*)malloc(sizeof(char*)*1000);
	novo.anoNascimento = (char*)malloc(sizeof(char*)*1000);
	novo.genero = (char*)malloc(sizeof(char*)*1000);
	novo.homeworld = (char*)malloc(sizeof(char*)*1000);
	return novo;
}

PERSONAGEM clone(PERSONAGEM personagem)
{
	PERSONAGEM clone;
	clone = personagem;
	return clone;
}

PERSONAGEM lerArquivo(char* filename, PERSONAGEM novo) //Le o arquivo e guarda tudo em uma variavel do struct
{
	if(filename[strlen(filename) - 1] == '\n')
	{
		filename[strlen(filename)-1] = '\0';	
	}
	FILE* arquivo = fopen(filename, "rt");
	char* buffer = (char*) malloc(sizeof(char*)*1000);
	char* tmp = (char*) malloc(sizeof(char*)*1000);


	do{
		fgets(buffer,300,arquivo);
	} while(!feof(arquivo) && strstr(buffer, "'films'") == NULL);

	novo.file = strdup(buffer);
	fclose(arquivo);
	return novo;
}

PERSONAGEM setNome(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");

	for(int x = 0; x < 2; x++)
	{
		tmp = strtok(NULL,"'");
	}
	novo.nome = strdup(tmp);
	return novo;
}

char* getNome(PERSONAGEM novo)
{
	return novo.nome;
}

PERSONAGEM setAltura(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");
	for(int x = 0; x < 6; x++)
	{
		tmp = strtok(NULL,"'");
	}
	novo.altura = atoi(tmp);
	return novo;
}

int getAltura(PERSONAGEM novo)
{
	return novo.altura;
}

PERSONAGEM setPeso(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");

	for(int x = 0; x < 10; x++)
	{
		tmp = strtok(NULL,"'");
	}
	if(strstr(tmp,"unknown"))
	{
		novo.peso = 0;
	}
	else{
		sscanf(tmp, "%lf", &novo.peso);
	}
	return novo;
}

double getPeso(PERSONAGEM novo)
{
	return novo.peso;
}

PERSONAGEM setCorDoCabelo(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");
	for(int x = 0; x < 14; x++)
	{
		tmp = strtok(NULL,"'");
	}
	novo.corDoCabelo = strdup(tmp);
	return novo;
}

char* getCorDoCabelo(PERSONAGEM novo)
{
	return novo.corDoCabelo;
}

PERSONAGEM setCorDaPele(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");
	for(int x = 0; x < 18; x++)
	{
		tmp = strtok(NULL,"'");
	}
	novo.corDaPele = strdup(tmp);
	return novo;
}

char* getCorDaPele(PERSONAGEM novo)
{
	return novo.corDaPele;
}

PERSONAGEM setCorDosOlhos(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");
	for(int x = 0; x < 22; x++)
	{
		tmp = strtok(NULL,"'");
	}
	novo.corDosOlhos = strdup(tmp);
	return novo;
}

char* getCorDosOlhos(PERSONAGEM novo)
{
	return novo.corDosOlhos;
}

PERSONAGEM setAnoNascimento(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");
	for(int x = 0; x < 26; x++)
	{
		tmp = strtok(NULL,"'");
	}
	novo.anoNascimento  = strdup(tmp);
	return novo;
}

char* getAnoNascimento(PERSONAGEM novo)
{
	return novo.anoNascimento;
}

PERSONAGEM setGenero(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");
	for(int x = 0; x < 30; x++)
	{
		tmp = strtok(NULL,"'");
	}
	novo.genero  = strdup(tmp);
	return novo;
}

char* getGenero(PERSONAGEM novo)
{
	return novo.genero;
}

PERSONAGEM setHomeworld(PERSONAGEM novo)
{
	char* cpy = strdup(novo.file);
	char* tmp = strtok(cpy,":");
	for(int x = 0; x < 34; x++)
	{
		tmp = strtok(NULL,"'");
	}
	novo.homeworld  = strdup(tmp);
	return novo;
}

char* getHomeworld(PERSONAGEM novo)
{
	return novo.homeworld;
}

PERSONAGEM setAll(PERSONAGEM novo)
{
	novo = setNome(novo);
	novo = setAltura(novo);
	novo = setPeso(novo);
	novo = setCorDoCabelo(novo);	
	novo = setCorDaPele(novo);	
	novo = setCorDosOlhos(novo);	
	novo = setAnoNascimento(novo);	
	novo = setGenero(novo);
	novo = setHomeworld(novo);
	return novo;
}

void printAll(PERSONAGEM novo) //metodo que imprime todos os gets
{	
	printf(" ## %s",novo.nome);
	printf(" ## %d",novo.altura);
	if((novo.peso - (int)novo.peso) == 0)
	{
		printf(" ## %.0lf",novo.peso);
	}
	else{
		printf(" ## %.1lf",novo.peso);
	}
	printf(" ## %s",novo.corDoCabelo);
	printf(" ## %s", novo.corDaPele);
	printf(" ## %s", novo.corDosOlhos);
	printf(" ## %s", novo.anoNascimento);
	printf(" ## %s", novo.genero);
	printf(" ## %s ## \n", novo.homeworld);
}

bool isFim(char* s){
	return (strlen(s) >= 3 && s[0] == 'F' && s[1] == 'I' && s[2] == 'M');
}


//=====================================================Fim Da Classe Personagem========================================================

typedef struct lista{
	PERSONAGEM array;
	int Tamanho;
}LISTA;

int quantidade = 0;
int tamanho = 100;
int numEntrada = 0;
int comp = 0;

void construirLista(char linha[][TAMLINHA], int numEntrada, LISTA lista[])
{
	for(int x = 0; x < numEntrada; x++)
	{
		lista[x].array = lerArquivo(linha[x],lista[x].array);
		lista[x].array = setAll(lista[x].array);
		quantidade++;
	}
}

void inserirInicio(LISTA lista[],PERSONAGEM personagem)
{
	if(quantidade < tamanho)
	{
		for(int x = quantidade; x > 0; x--)
		{
			lista[x].array = clone(lista[x-1].array);
		}
		lista[0].array = clone(personagem);
		quantidade++;
	}
	else{
		printf("\nO array esta cheio");
	}
}

void inserir(LISTA lista[],PERSONAGEM personagem,int pos)
{
	if(quantidade < tamanho)
	{
		for(int x = quantidade; x > pos; x--)
		{
			lista[x].array = clone(lista[x-1].array);
		}
		lista[pos].array = clone(personagem);
		quantidade++;
	}
	else{
		printf("\nO array esta cheio");
	}
}

void inserirFim(LISTA lista[],PERSONAGEM personagem)
{
	if(quantidade < tamanho)
	{
		lista[quantidade].array = clone(personagem);
		quantidade++;
	}
	else{
		printf("\nO array esta cheio");
	}
}

PERSONAGEM removerInicio(LISTA lista[])
{
	PERSONAGEM resp = clone(lista[0].array);
	setNome(resp);

	if(quantidade != 0)
	{
		quantidade--;
		for(int x = 0; x < quantidade; x++)
		{
			lista[x].array = clone(lista[x+1].array);
		}
	}
	return resp;
}

PERSONAGEM remover(LISTA lista[], int pos)
{
	PERSONAGEM resp = clone(lista[pos].array);

	if(quantidade > 0 && pos >= 0 && pos < quantidade)
	{
		quantidade--;

		for(int x = pos; x < quantidade; x++)
		{
			lista[x].array = clone(lista[x+1].array);
		}
		setNome(resp);
	}
	return resp;
}

PERSONAGEM removerFim(LISTA lista[])
{
	PERSONAGEM resp = clone(lista[quantidade].array);

	if(quantidade > 0)
	{
		quantidade--;
		lista[quantidade].array = setNome(lista[quantidade].array);
	}
	return lista[quantidade].array;
}

/*void sort(LISTA lista[],int i, int j, int size, int flag )
{
	PERSONAGEM tmp;

	if( i < size)
	{
		if(flag)
		{
			j = i + 1;
		}
		if(j < size)
		{
			if(strcmp( getNome(lista[i].array), getNome(lista[j].array)) > 0 )
			{
				tmp = clone(lista[i].array);
				lista[i].array = clone(lista[j].array);
				lista[j].array = clone(tmp);
			}
			sort(lista, i, j + 1, size, 0);
		}
		sort(lista, i + 1, 0, size, 1);
	}
}*/



void mostrarArray(LISTA lista[])
{
	for(int x = 0; x < quantidade; x++)
	{
		lista[x].array = setAll(lista[x].array);
	}

	for(int x = 0; x < quantidade; x++)
	{
		printf("[%d] ",x);
		printAll(lista[x].array);
	}
}

int main(int argc, char** argv)
{
	char entrada[NUMENTRADA][TAMLINHA];
	char entrada2[NUMENTRADA][TAMLINHA];
	char* nome = (char*) malloc(sizeof(char*)*1000);
	int numEntrada2 = 0;

	clock_t startTime = clock();

	//Ler todas as entradas
	do {
		fgets(entrada[numEntrada], TAMLINHA, stdin);
	} while (isFim(entrada[numEntrada++]) == false);
	numEntrada--;

	PERSONAGEM personagem[tamanho];
	LISTA lista[tamanho];
	construirLista(entrada,numEntrada,lista);

	sort(lista, 0 , 0 , quantidade, 1);
	mostrarArray(lista);
}
